File: ds/C-C++/graphs/bfs.cpp
#include <iostream>
#include <queue>
using namespace std;

typedef struct Node {
    Node* left;
    Node* right;
    int data;
} Node ;

void BFS(Node* root) {
    if (root == NULL) return;

    queue<Node*> q;
    q.push(root);

    while (!q.empty()) {
        Node* curr = q.front();
        cout << curr->data << " ";
        if (curr->left != NULL) q.push(curr->left);
        if(curr->right != NULL) q.push(curr->right);
        q.pop();
    }
}

int main (void)
{
    Node* root = new Node();
    root->data = 5;
    root->left = new Node();
    root->right = new Node();

    root->left->data = 7;
    root->right->data = 6;

    BFS(root);
    free(root);
    return 0;
}

File: ds/C/Tree/BinaryTree/bfs.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


# Time: O(n) Space: O(n)
def levelOrder(root: Node):
    if root is None:
        return

    discovered = [root]
    while len(discovered) > 0:
        node = discovered.pop(0)
        print(node.data, end=" ")  # visiting the node
        if node.left:
            discovered.append(node.left)
        if node.right:
            discovered.append(node.right)


root = Node('F')
root.left = Node('D')
root.right = Node('J')
root.left.left = Node('B')
root.left.right = Node('E')
root.left.left.left = Node('A')
root.left.left.right = Node('C')
root.right.left = Node('G')
root.right.right = Node('K')
root.right.left.right = Node('I')
root.right.left.right.left = Node('H')

levelOrder(root)
print()
